org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/SourceTypeBinding.java

```diff
@@ -3263,9 +3228,31 @@ public ReferenceBinding setSuperClass(ReferenceBinding superClass) {
 			annotatedType.superclass = superClass;
 		}
 	}
+	if (superClass != null && superClass.actualType() instanceof SourceTypeBinding sourceSuperType && sourceSuperType.isSealed()
+			&& (sourceSuperType.scope.referenceContext.permittedTypes == null || (sourceSuperType.scope.referenceContext.permittedTypes.length > 0 && sourceSuperType.scope.referenceContext.permittedTypes[0].isImplicit()))) {
+		sourceSuperType.setImplicitPermittedType(this);
+		if (this.isAnonymousType() && superClass.isEnum())
+			this.modifiers |= ClassFileConstants.AccFinal;
+	}
 	return this.superclass = superClass;
 }
 
+private void setImplicitPermittedType(SourceTypeBinding permittedType) {
+	ReferenceBinding[] typesPermitted = this.permittedTypes();
+	int sz = typesPermitted == null ? 0 : typesPermitted.length;
+	if (this.scope.referenceCompilationUnit() == permittedType.scope.referenceCompilationUnit()) {
+		if (sz == 0) {
+			typesPermitted = new ReferenceBinding[] { permittedType };
+		} else {
+			System.arraycopy(typesPermitted, 0, typesPermitted = new ReferenceBinding[sz + 1], 0, sz);
```

Comment:
> I needed to nitpick and here it is: You are guaranteed n-1 arraycopy() calls where n = SuperType + SuperInterfaces defined in the same compilation unit. Any chances of using List<> and converting to array once all the additions are done.

Thanks, I will make a note of it and see if it can be included - there is several more incremental fixes/cleanups coming